### ISSUE TITLE
Referring to anonymous type parameters (captured via 'any') of generics leads to silent any coercition.

### ISSUE THREAD

**Issue by sledorze** [Microsoft/TypeScript#22025]
**TypeScript Version:**  2.7.2 and 2.8.0-dev.20180217

Search Terms: any, capture, 

**Code**

```ts
interface Foo<T> {
  t: T
}
type TFoo<F extends Foo<any>> = F['t']

const bar = <Q extends Foo<any>>(f: Q): TFoo<Q> => {
  return f.t
}

const bar2 = <Q extends Foo<any>>(f: Q): TFoo<Q> => {
  return [f.t] // wrong type not caught by compiler
}

const foo1: Foo<number> = {
  t: 45
}

const foo2: Foo<number[]> = {
  t: [45]
}

bar(foo1) // number
bar(foo2) // number[]

bar2(foo1) // number !!!  (but in reality it returns a number[])
bar2(foo2) // number[] !!!  (but in reality it returns a number[][])
Webpack, etc.
```

**Expected behavior:**
I KNOW that from the types, this behaviour might be expected.
However, I think that the pattern exhibited here is a common one and the only one I know of to be able to abstract over generic types **without** having to specify those types explicitely (which leads to other variance issues in specific cases).

I would have loved to have an error when implementing bar2, like it does if I do:

```ts
const bar2 = <X, Q extends Foo<T>>(f: Q): TFoo<Q> => {
  return [f.t] // Type 'X[]' is not assignable to type 'Q["t"]'.  Type 'X[]' is not assignable to type 'X'.
}
```
it may be:

```ts
const bar2 = <Q extends Foo<any>>(f: Q): TFoo<Q> => {
  return [f.t] // Type 'Q["t"][]' is not assignable to type 'Q["t"]'.
}
```

**Actual behavior:**

There's no compiler error

**Related Issues:**

https://github.com/gcanti/io-ts-types/issues/18

**Comment by uhyo** [comment]
Conditional types provide a way to introduce a type variable in a type definition. So a workaround would be:

```ts
const bar2 = <Q extends Foo<any>>(f: Q): Q extends Foo<infer U> ? TFoo<Foo<U>> : never => {
 // error TS2322: Type 'any[]' is not assignable to type 'Q extends Foo<U> ? U : never'.
  return [f.t]
}
```

**Comment by sledorze** [comment]
@uhyo thanks for the info, it can be helpful.

Back to the original issue, you can fix it when you know it.

The main problem here is that it's a silent error; the compiler is accepting a wrong program with all strict flags activated and there's no way to detect that kind of error in a code base.

**Comment by jack-williams** [comment]
Does this work?

```ts
const bar2 = <Q extends Foo<W>, W = any>(f: Q): TFoo<Q> => {
    return [f.t] // wrong type not caught by compiler
}
```

**Comment by sledorze** [comment]
@jack-williams indeed, this workaround looks to behave correctly.

**Comment by sledorze** [comment]
The explanation is that:

- in the function, the compiler cannot infer to any as there's a type param that can be bound (W).
- out of the function, it correctly infers from the context, extracting the real types via the structure (TFoo).

Beware, one should not directly use W in the Result type.

It look like a warn/error when a Type param bound to `any` appears in the Result of a function would detect much of those cases (not sure if it would break a lot of 'valid' code).

**Comment by mhegazy** [comment]
Please see https://github.com/Microsoft/TypeScript/issues/19412
